File: internal/2024/day4/go/main.go
package main

import (
	"advent/internal/pkg/reader"
	"fmt"
)

func main() {
	data := reader.FileTo2DArray("data/2024/day4.txt")

	count := 0
	for y := range data {
		for x := range data[y] {
			count += xmasCount(data, y, x)
		}
	}
	fmt.Println("XMAS count:", count)

	count = 0
	for y := 0; y < len(data)-2; y++ {
		for x := 0; x < len(data[y])-2; x++ {
			temp := make([][]rune, 3)
			for i := 0; i < 3; i++ {
				temp[i] = data[y+i][x:x+3]
			}
			if hasXMas(temp) {
				count++
			}
		}
	}
	fmt.Println("X-MAS count:", count)
}

const (
	NORTH = "north"
	NORTH_EAST = "north east"
	EAST = "east"
	SOUTH_EAST = "south east"
	SOUTH = "south"
	SOUTH_WEST = "south west"
	WEST = "west"
	NORTH_WEST = "north west"
)

var cardinalities = map[string][2]int{
	NORTH: [2]int{-1, 0},
	NORTH_EAST: [2]int{-1, 1},
	EAST: [2]int{0, 1},
	SOUTH_EAST: [2]int{1, 1},
	SOUTH: [2]int{1, 0},
	SOUTH_WEST: [2]int{1, -1},
	WEST: [2]int{0, -1},
	NORTH_WEST: [2]int{-1, -1},
}

func xmasCount(data [][]rune, y, x int) int {
	result := 0
	target := []rune("XMAS")
	loop:
	for _, card := range cardinalities {
		for i := 0; i < len(target); i++ {
			posY := y + card[0] * i 
			posX := x + card[1] * i
			if posY < 0 || posY >= len(data) {
				continue loop
			}
			if posX < 0 || posX >= len(data[posY]) {
				continue loop
			}
			if data[posY][posX] != target[i] {
				continue loop
			}
		}
		result++
	}
	return result
}

func hasXMas(data [][]rune) bool {
	if data[1][1] != 'A' {
		return false 
	}

	target := "MS"
	tr := string(data[0][0]) + string(data[2][2])
	rt := string(data[2][2]) + string(data[0][0])
	if target != tr && target != rt {
		return false
	}

	br := string(data[2][0]) + string(data[0][2])
	rb := string(data[0][2]) + string(data[2][0])
	return target == br || target == rb
}

File: internal/2024/day4/cpp/main.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

int countXmasses(std::vector<std::vector<char>>, int, int);
bool hasXMas(std::vector<std::vector<char>>);

const int cardinalities[8][2] = {{-1, -1}, {-1,0}, {-1, 1}, {0, 1}, {1, 1}, {1, 0}, {1, -1}, {0, -1}};

int main() {
	std::string file_path = "data/2024/day4.txt";

	std::ifstream file(file_path);
	if (!file.is_open()) {
		std::cerr << "Error: could not open file " << file_path << std::endl;
		return 1;
	}

	std::string line;
	std::vector<std::vector<char>> data;
	while (std::getline(file, line)) {
		std::vector<char> row;
		for (int i = 0; i < line.length(); i++) {
			row.push_back(line[i]);
		}
		data.push_back(row);
	}

	int xmasses = 0;
	for (int i = 0; i < data.size(); i++) {
		for (int j = 0; j < data[i].size(); j++) {
			xmasses += countXmasses(data, i, j);
		}
	}
	std::cout << "XMAS count: " << xmasses << std::endl;

	int xMasses = 0;
	for (int i = 0; i < data.size()-2; i++) {
		for (int j = 0; j < data[i].size()-2; j++) {
			std::vector<std::vector<char>> temp;
			for (int r = 0; r < 3; r++) {
				std::vector<char> temp_row = {data[i+r][j], data[i+r][j+1], data[i+r][j+2]};
				temp.push_back(temp_row);
			}
			if (hasXMas(temp)) {
				xMasses++;
			}
		}
	}

	std::cout << "X-MAS count: " << xMasses << std::endl;
	return 0;
}

bool hasXMas(std::vector<std::vector<char>> data) {
	if (data[1][1] != 'A') {
		return false;
	}
	std::string target = "MS";

	std::string tr = std::string(1, data[2][2]) + std::string(1, data[0][0]);
	std::string rt = std::string(1, data[0][0]) + std::string(1, data[2][2]);
	if (tr != target && rt != target) {
		return false;
	}

	std::string br = std::string(1, data[2][0]) + std::string(1, data[0][2]);
	std::string rb = std::string(1, data[0][2]) + std::string(1, data[2][0]);
	return br == target || rb == target;
}

int countXmasses(std::vector<std::vector<char>> data, int y, int x) {
	int count = 0;
	std::string target = "XMAS";
	bool found = false;
	for (auto card:cardinalities) {
		found = true;
		for (int i = 0; i < target.length(); i++) {
			int posY = y + card[0] * i;
			int posX = x + card[1] * i;
			if (posY < 0 || posY >= data.size()) {
				found = false;
				break;
			}
			if (posX < 0 || posX >= data[posY].size()) {
				found = false;
				break;
			}
			if (data[posY][posX] != target[i]) {
				found = false;
				break;
			}
		}
		if (found) {
			count++;
		}
	}
	return count;
}
